feat: build and publish binaries on release by using make and action-gh-release

kurokobo opened the pull request:
Closes #719 

@TheRealHaoLiu 
I saw that you were already assigned to the issue, so feel free to ignore this PR, but I've created this as a working demo 😃 

### Design

In this PR, I use `make` and `action-gh-release` to build and publish binaries.

- **`make` and `action-gh-release`**
  - **Pros:**
    - Fully customizable
    - Any tags are allowed even if the tag does not satisfy semver
  - **Cons:**
    - Building rpms and debs can be a bit more complicated

- **GoReleaser**
  - **Pros:**
    - Supports building rpms and debs in addition to building binaries
  - **Cons:**
    - Less customizable especially for rpms and debs
    - The tags for the release MUST satisfy semver (current `v1.3.0.dev1` is not allowed)

The pros for GoReleaser is that it also supports building RPMs in addition to just binaries, but compiling SELinux policies and generating Bash Completion are required before GoReleaer, so it ends up being a bit more complicated.

Also, since non-semver tags are not allowed, the workflow will fail if we pass the tags like `v1.3.0.dev1` when releasing manually. 

So, first, I've created a PR to publish binaries using `make` in a straightforward and simple manner.

For demo purpose, GoReleaser version is here: https://github.com/ansible/receptor/pull/725

### Test

Demo release is available on my forked repository: <https://github.com/kurokobo/receptor/releases/tag/v1.3.1-demo1a>

1. Create release by hand
   ![image](https://user-images.githubusercontent.com/2920259/216241064-830587ac-9124-445c-a2f9-9429be247712.png)
2. Wait for the the job `publish` in `promote.yml` to be completed.
   - <https://github.com/kurokobo/receptor/actions/runs/4071606857/jobs/7013552996>
3. Ensure the release has been updated and binaries are published.
   ![image](https://user-images.githubusercontent.com/2920259/216241992-1d8e7788-fab3-442a-8452-ec99d775e0ae.png)
4. Download binaries and ensure works well.
   ```bash
   $ curl -LO https://github.com/kurokobo/receptor/releases/download/v1.3.1-demo1a/receptor_1.3.1-demo1a_linux_amd64.tar.gz
     % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                    Dload  Upload   Total   Spent    Left  Speed
     0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
   100 20.8M  100 20.8M    0     0   9.8M      0  0:00:02  0:00:02 --:--:-- 18.4M
   
   $ tar zxvf receptor_1.3.1-demo1a_linux_amd64.tar.gz 
   receptor
   
   $ ./receptor --version
   v1.3.1-demo1a
   ```
   ```powershell
   PS C:\Temp> Invoke-WebRequest https://github.com/kurokobo/receptor/releases/download/v1.3.1-demo1a/receptor_1.3.1-demo1a_windows_amd64.tar.gz -OutFile receptor_1.3.1-demo1a_windows_amd64.tar.gz
   
   PS C:\Temp> tar zxvf .\receptor_1.3.1-demo1a_windows_amd64.tar.gz
   x receptor.exe
   
   PS C:\Temp> .\receptor.exe --version
   v1.3.1-demo1a
   ```

kurokobo committed:
feat: build and publish binaries on release by using make and action-gh-release

TheRealHaoLiu commented:
@kurokobo ur the coolest ❤️

kurokobo commented:
Thanks for moving forward this PR.
I've updated the code to remove top-level subdirectory inside the tar.gz, and also updated the above PR comment.

tanganellilore commented:
team, some news on it?

AaronH88 commented on Makefile:
```diff
@@ -71,6 +71,14 @@ build-all:
 	go build -o receptor ./cmd/receptor-cl
 
+DIST := receptor_$(shell echo '$(VERSION)' | sed 's/^v//')_$(GOOS)_$(GOARCH)
+build-package:
```
Looks really good, 
Can you add cleaning the `dist` dir to the `clean:` make target please?

kurokobo replied:
Thanks, I totally forgot about `clean`. Added.

AaronH88 commented:
This looks great!
You have got a really good explanation on how to use this release process in the PR, can this explanation be added to the receptor docs please? Perhaps `https://github.com/ansible/receptor/blob/devel/docs/source/release_process.rst` is a good place to update the docs

kurokobo commented:
@AaronH88 
Thanks for your review! I've modified, rebased, and pushed to solve your comments.

As for the documentation, I just wrote an overview without mentioning specific tools, what do you think?
![image](https://user-images.githubusercontent.com/2920259/223431756-79f3bf7c-4b81-46a5-ad22-940cfc1a3808.png)

kurokobo commented:
I forgot to mention that we need to change the permissions for GitHub Actions to `Read and write permissions` in the project settings (`Settings` > `Actions` > `General`) for this workflow to work.

![image](https://user-images.githubusercontent.com/2920259/223432013-bde8494a-d63f-4145-9c9a-94d9792a8825.png)

AaronH88 commented:
great work @kurokobo , running the github actions now :smile:

AaronH88 commented:
@shanemcd this PR linting is throwing a false positive.
I have reviewed the code and feel its ok to merge, however, I can not see the permissions for the workflows as outlined in the comments. Can you see these permissions?

AaronH88 commented:
@kurokobo I have just checked with the internal team and the action permissions are set to "Read and Write",
Can you do a quick rebase please so we can get rid of this linting error?

kurokobo commented:
@AaronH88 
Thanks for your confirmation, I've just rebased and force-pushed to my branch.

AaronH88 reviewed (approved):
/lgtm
